File: src/#784/solution.js
var letterCasePermutation = function (S) {
  let output = []
  helper(0, '')

  return output

  function helper(index, currStr) {
    if (index >= S.length) {
      output.push(currStr)
      return
    }
    let code = S[index].charCodeAt()
    if (code >= 48 && code <= 57) {
      helper(index + 1, currStr + S[index])
      return
    }
    if (code >= 65 && code <= 90) {
      helper(index + 1, currStr + S[index])
      helper(index + 1, currStr + String.fromCharCode(code + 32))
    }
    if (code >= 97 && code <= 122) {
      helper(index + 1, currStr + S[index])
      helper(index + 1, currStr + String.fromCharCode(code - 32))
    }
  }
}

File: src/#784/solution.go
package solution

func letterCasePermutation(S string) []string {
	bytes := []byte(S)
	output := []string{}
	helper(0, bytes, []byte{}, &output)
	return output
}

func helper(index int, str []byte, currStr []byte, output *[]string) {
	if index >= len(str) {
		*output = append(*output, string(currStr))
		return
	}

	if str[index] <= '9' && str[index] >= '1' {
		helper(index+1, str, append(currStr, str[index]), output)
		return
	}
	if str[index] >= 'A' && str[index] <= 'Z' {
		helper(index+1, str, append(currStr, str[index]), output)
		helper(index+1, str, append(currStr, str[index]+32), output)
		return
	}
	if str[index] >= 'a' && str[index] <= 'z' {
		helper(index+1, str, append(currStr, str[index]), output)
		helper(index+1, str, append(currStr, str[index]-32), output)
		return
	}
}
